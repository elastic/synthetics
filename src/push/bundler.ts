/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import path from 'path';
import { stat, unlink, readFile } from 'fs/promises';
import { createWriteStream } from 'fs';
import * as esbuild from 'esbuild';
import archiver from 'archiver';
import { commonOptions } from '../core/transform';
import { SyntheticsBundlePlugin } from './plugin';

const SIZE_LIMIT_KB = 800;

function relativeToCwd(entry: string) {
  return path.relative(process.cwd(), entry);
}

export class Bundler {
  moduleMap = new Map<string, string>();
  constructor() {}

  async prepare(absPath: string) {
    const options: esbuild.BuildOptions = {
      ...commonOptions(),
      ...{
        entryPoints: [absPath],
        bundle: true,
        write: false,
        sourcemap: 'inline',
        external: ['@elastic/synthetics'],
        plugins: [SyntheticsBundlePlugin()],
      },
    };
    const result = await esbuild.build(options);
    if (result.errors.length > 0) {
      throw result.errors;
    }
    this.moduleMap.set(absPath, result.outputFiles[0].text);
  }

  async zip(outputPath: string) {
    return new Promise((fulfill, reject) => {
      const output = createWriteStream(outputPath);
      const archive = archiver('zip', {
        zlib: { level: 9 },
      });
      archive.on('error', reject);
      output.on('close', fulfill);
      archive.pipe(output);
      for (const [path, content] of this.moduleMap.entries()) {
        const relativePath = relativeToCwd(path);
        // Date is fixed to Unix epoch so the file metadata is
        // not modified everytime when files are bundled
        archive.append(content, {
          name: relativePath,
          date: new Date('1970-01-01'),
        });
      }
      archive.finalize();
    });
  }

  async build(entry: string, output: string) {
    await this.prepare(entry);
    await this.zip(output);
    const data = await this.encode(output);
    await this.checkSize(output);
    await this.cleanup(output);
    return data;
  }

  async encode(outputPath: string) {
    return await readFile(outputPath, 'base64');
  }

  async checkSize(outputPath: string) {
    const { size } = await stat(outputPath);
    const sizeKb = size / 1024;
    if (sizeKb > SIZE_LIMIT_KB) {
      throw new Error(
        `You have monitors whose size exceeds the ${SIZE_LIMIT_KB}KB limit.`
      );
    }
  }

  async cleanup(outputPath: string) {
    this.moduleMap = new Map<string, string>();
    await unlink(outputPath);
  }
}
