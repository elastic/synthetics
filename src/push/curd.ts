/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { HttpMethod } from 'undici/types/dispatcher';
import { PushOptions } from '../common_types';
import { removeTrailingSlash } from '../helpers';
import {
  formatAPIError,
  formatNotFoundError,
  ok,
  sendRequest,
} from './request';

export type MonitorData = {
  journey_id?: string;
  hash?: string;
};

type GetResponse = {
  total: number;
  monitors: MonitorData[];
};

export function generateURL(options: PushOptions, method: HttpMethod) {
  switch (method) {
    case 'GET':
      return (
        removeTrailingSlash(options.url) +
        `/s/${options.space}/api/synthetics/project/${options.id}/monitors`
      );
    case 'PUT':
      return '';
    case 'DELETE':
      return '';
  }
  throw 'Method not supported';
}

export async function getAllMonitors(options: PushOptions) {
  const { body, statusCode } = await sendRequest({
    url: generateURL(options, 'GET'),
    method: 'GET',
    auth: options.auth,
  });

  if (statusCode === 404) {
    throw formatNotFoundError(await body.text());
  }
  if (!ok(statusCode)) {
    const { error, message } = await body.json();
    throw formatAPIError(statusCode, error, message);
  }

  const { monitors } = (await body.json()) as GetResponse;
  return monitors;
}
