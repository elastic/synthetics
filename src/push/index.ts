/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
import { readFile, writeFile } from 'fs/promises';
import { prompt } from 'enquirer';
import { bold, grey } from 'kleur/colors';
import {
  getLocalMonitors,
  buildMonitorSchema,
  diffMonitors as diffMonitorHashIDs,
  MonitorSchema,
} from './monitor';
import { ALLOWED_SCHEDULES, Monitor } from '../dsl/monitor';
import {
  progress,
  liveProgress,
  write,
  error,
  warn,
  indent,
  done,
  getMonitorManagementURL,
  THROTTLING_WARNING_MSG,
} from '../helpers';
import type { PushOptions, ProjectSettings } from '../common_types';
import { findSyntheticsConfig, readConfig } from '../config';
import {
  bulkDeleteMonitors,
  bulkGetMonitors,
  bulkPutMonitors,
  createMonitorsLegacy,
  CHUNK_SIZE,
} from './kibana_api';
import {
  getChunks,
  isBulkAPISupported,
  isLightweightMonitorSupported,
  logDiff,
} from './utils';

export async function push(monitors: Monitor[], options: PushOptions) {
  const duplicates = trackDuplicates(monitors);
  if (duplicates.size > 0) {
    throw error(formatDuplicateError(duplicates));
  }
  progress(
    `Pushing monitors for '${options.id}' project in kibana '${options.space}' space`
  );

  /**
   * Legacy API for kibana which does not support bulk operations
   */
  if (!isBulkAPISupported(options.kibanaVersion)) {
    return await pushLegacy(monitors, options);
  }

  const { monitors: remote } = await bulkGetMonitors(options);

  progress(`bundling ${monitors.length} monitors`);
  const schemas = await buildMonitorSchema(monitors, true);
  const local = getLocalMonitors(schemas);

  const { newIDs, changedIDs, removedIDs, unchangedIDs } = diffMonitorHashIDs(
    local,
    remote
  );
  logDiff(newIDs, changedIDs, removedIDs, unchangedIDs);

  const updatedMonitors = new Set<string>([...changedIDs, ...newIDs]);
  if (updatedMonitors.size > 0) {
    const chunks = getChunks(schemas, CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkPutMonitors(options, chunk),
        `creating or updating ${chunk.length} monitors`
      );
    }
  }

  if (removedIDs.size > 0) {
    if (updatedMonitors.size === 0 && unchangedIDs.size === 0) {
      await promptConfirmDeleteAll(options);
    }
    const chunks = getChunks(Array.from(removedIDs), CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkDeleteMonitors(options, chunk),
        `deleting ${chunk.length} monitors`
      );
    }
  }

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

async function promptConfirmDeleteAll(options: PushOptions) {
  write('');
  const { deleteAll } = await prompt<{ deleteAll: boolean }>({
    type: 'confirm',
    skip() {
      if (options.yes) {
        this.initial = process.env.TEST_OVERRIDE ?? true;
        return true;
      }
      return false;
    },
    name: 'deleteAll',
    message: `Pushing without any monitors will delete all monitors associated with the project.\n Do you want to continue?`,
    initial: false,
  });
  if (!deleteAll) {
    throw warn('Push command Aborted');
  }
}

function trackDuplicates(monitors: Monitor[]) {
  const monitorMap: Map<string, Monitor> = new Map();
  const duplicates: Set<Monitor> = new Set();
  for (const monitor of monitors) {
    const id = monitor.config.id;
    if (monitorMap.has(id)) {
      duplicates.add(monitorMap.get(id));
      duplicates.add(monitor);
    }
    monitorMap.set(id, monitor);
  }
  return duplicates;
}

export function formatDuplicateError(monitors: Set<Monitor>) {
  let outer = bold(`Aborted: Duplicate monitors found\n`);

  let inner = '';
  for (const monitor of monitors) {
    const { config, source } = monitor;
    inner += `* ${config.id} - ${source.file}:${source.line}:${source.column}\n`;
  }
  outer += indent(inner);
  return outer;
}

const INSTALLATION_HELP = `Run 'npx @elastic/synthetics init' to create project with default settings.`;

export async function loadSettings(configPath, ignoreMissing = false) {
  try {
    const config = await readConfig(
      process.env['NODE_ENV'] || 'development',
      configPath
    );
    // Missing config file, fake throw to capture as missing file
    if (Object.keys(config).length === 0) {
      throw '';
    }
    return config.project || ({} as ProjectSettings);
  } catch (e) {
    if (!ignoreMissing) {
      throw error(`Aborted (missing synthetics config file), Project not set up correctly.

${INSTALLATION_HELP}`);
    }
  }
}

export async function validatePush(
  opts: PushOptions,
  settings: ProjectSettings
) {
  validateSettings(opts);
  await catchIncorrectSettings(settings, opts);
}

export function validateSettings(opts: PushOptions) {
  const INVALID = 'Aborted. Invalid synthetics project settings.';

  let reason = '';
  if (!opts.id) {
    reason = `Set project id via
  - CLI '--id <id>'
  - Config file 'project.id' field`;
  } else if (!opts.locations && !opts.privateLocations) {
    reason = `Set default location for all monitors via
  - CLI '--locations <values...> or --privateLocations <values...>'
  - Config file 'monitors.locations' | 'monitors.privateLocations' field`;
  } else if (!opts.schedule) {
    reason = `Set default schedule in minutes for all monitors via
  - CLI '--schedule <mins>'
  - Config file 'monitors.schedule' field`;
  } else if (opts.schedule && !ALLOWED_SCHEDULES.includes(opts.schedule)) {
    reason = `Set default schedule(${
      opts.schedule
    }) to one of the allowed values - ${ALLOWED_SCHEDULES.join(',')}`;
  }

  if (!reason) return;

  throw error(`${INVALID}

${reason}

${INSTALLATION_HELP}`);
}

async function overrideSettings(
  configPath,
  oldValue: string,
  newValue: string
) {
  const cwd = process.cwd();
  configPath = configPath ?? (await findSyntheticsConfig(cwd, cwd));
  if (!configPath) {
    throw warn(`Unable to find synthetics config file: ${configPath}`);
  }
  const config = await readFile(configPath, 'utf-8');
  const updatedConfig = config.replace(
    `id: '${oldValue}'`,
    `id: '${newValue}'`
  );
  await writeFile(configPath, updatedConfig, 'utf-8');
}

export async function catchIncorrectSettings(
  settings: ProjectSettings,
  options: PushOptions
) {
  let override = !settings.id;
  if (settings.id && settings.id !== options.id) {
    // Add an extra line to make it easier to read the prompt
    write('');
    ({ override } = await prompt<{ override: boolean }>({
      type: 'confirm',
      name: 'override',
      skip() {
        if (options.yes) {
          this.initial = process.env.TEST_OVERRIDE ?? true;
          return true;
        }
        return false;
      },
      message: `Monitors were pushed under the '${settings.id}' project. Are you sure you want to push them under the new '${options.id}' (note that this will duplicate the monitors, the old ones being orphaned)`,
      initial: false,
    }));
    if (!override) {
      throw warn('Push command Aborted');
    }
  }
  if (override) {
    await overrideSettings(options.config, settings.id, options.id);
  }
}

export async function pushLegacy(monitors: Monitor[], options: PushOptions) {
  if (isLightweightMonitorSupported(monitors, options)) {
    throw error(
      `Aborted: Lightweight monitors are not supported in ${options.kibanaVersion}. Please upgrade to 8.5.0 or above.`
    );
  }

  let schemas: MonitorSchema[] = [];
  if (monitors.length > 0) {
    progress(`bundling ${monitors.length} monitors`);
    schemas = await buildMonitorSchema(monitors, false);
    const chunks = getChunks(schemas, 10);
    for (const chunk of chunks) {
      await liveProgress(
        createMonitorsLegacy({ schemas: chunk, keepStale: true, options }),
        `creating or updating ${chunk.length} monitors`
      );
    }
  } else {
    await promptConfirmDeleteAll(options);
  }
  await liveProgress(
    createMonitorsLegacy({ schemas, keepStale: false, options }),
    `deleting all stale monitors`
  );

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

// prints warning if any of the monitors has throttling settings enabled during push
export function warnIfThrottled(monitors: Monitor[]) {
  const throttled = monitors.some(monitor => monitor.config.throttling != null);
  if (throttled) {
    warn(THROTTLING_WARNING_MSG);
  }
}
