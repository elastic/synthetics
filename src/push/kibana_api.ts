/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { PushOptions } from '../common_types';
import { safeNDJSONParse } from '../helpers';
import { MonitorHashID, MonitorSchema } from './monitor';
import {
  formatFailedMonitors,
  formatStaleMonitors,
  handleError,
  sendReqAndHandleError,
  sendRequest,
  APIMonitorError,
} from './request';
import { generateURL } from './utils';

// Default chunk size for bulk put / delete
export const CHUNK_SIZE = 100;

export type PutResponse = {
  createdMonitors: string[];
  updatedMonitors: string[];
  failedMonitors: APIMonitorError[];
};

export async function bulkPutMonitors(
  options: PushOptions,
  schemas: MonitorSchema[]
) {
  const resp = await sendReqAndHandleError<PutResponse>({
    url: generateURL(options, 'bulk_update') + '/_bulk_update',
    method: 'PUT',
    auth: options.auth,
    body: JSON.stringify({ monitors: schemas }),
  });
  const { failedMonitors } = resp;
  if (failedMonitors && failedMonitors.length > 0) {
    throw formatFailedMonitors(failedMonitors);
  }
  return resp;
}

export type GetResponse = {
  total: number;
  monitors: MonitorHashID[];
  after_key?: string;
};

export async function bulkGetMonitors(
  options: PushOptions
): Promise<GetResponse> {
  const allMonitors: MonitorHashID[] = [];

  const resp = await fetchMonitors(options);
  allMonitors.push(...resp.monitors);
  let afterKey = resp.afterKey;
  const total = resp.total;

  while (allMonitors.length < total) {
    const resp = await fetchMonitors(options, afterKey);
    allMonitors.push(...resp.monitors);
    afterKey = resp.afterKey;
  }

  return {
    total,
    monitors: allMonitors,
  };
}

const fetchMonitors = async (options: PushOptions, afterKey?: string) => {
  let url = generateURL(options, 'bulk_get');
  if (afterKey) {
    url += `?search_after=${afterKey}`;
  }
  const resp = await sendReqAndHandleError<GetResponse>({
    url,
    method: 'GET',
    auth: options.auth,
  });
  return {
    afterKey: resp.after_key,
    total: resp.total,
    monitors: resp.monitors,
  };
};

export type DeleteResponse = {
  deleted_monitors: string[];
};

export async function bulkDeleteMonitors(
  options: PushOptions,
  monitorIDs: string[]
) {
  return await sendReqAndHandleError<DeleteResponse>({
    url: generateURL(options, 'bulk_delete') + '/_bulk_delete',
    method: 'DELETE',
    auth: options.auth,
    body: JSON.stringify({ monitors: monitorIDs }),
  });
}

type StatusResponse = {
  kibana: {
    version: string;
  };
};

export async function getVersion(options: PushOptions) {
  const data = await sendReqAndHandleError<StatusResponse>({
    url: generateURL(options, 'status'),
    method: 'GET',
    auth: options.auth,
  });

  return data.kibana.version;
}

export type LegacyAPISchema = {
  project: string;
  keep_stale: boolean;
  monitors: MonitorSchema[];
};

export async function createMonitorsLegacy({
  schemas,
  keepStale,
  options,
}: {
  schemas: MonitorSchema[];
  keepStale: boolean;
  options: PushOptions;
}) {
  const schema: LegacyAPISchema = {
    project: options.id,
    keep_stale: keepStale,
    monitors: schemas,
  };
  const url = generateURL(options, 'legacy');
  const { body, statusCode } = await sendRequest({
    url,
    method: 'PUT',
    auth: options.auth,
    body: JSON.stringify(schema),
  });

  const resBody = await handleError(statusCode, url, body);
  const allchunks = [];
  for await (const data of resBody) {
    allchunks.push(Buffer.from(data));
  }
  const chunks = safeNDJSONParse(Buffer.concat(allchunks).toString('utf-8'));
  // Its kind of hacky for now where Kibana streams the response by
  // writing the data as NDJSON events (data can be interleaved), we
  // distinguish the final data by checking if the event was a progress vs complete event
  for (const chunk of chunks) {
    if (typeof chunk === 'string') {
      // Ignore the progress from Kibana as we chunk the requests
      continue;
    }
    const { failedMonitors, failedStaleMonitors } = chunk;
    if (failedMonitors && failedMonitors.length > 0) {
      throw formatFailedMonitors(failedMonitors);
    }
    if (failedStaleMonitors.length > 0) {
      throw formatStaleMonitors(failedStaleMonitors);
    }
  }
}
