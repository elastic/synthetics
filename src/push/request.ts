/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { bold } from 'kleur/colors';
import { request } from 'undici';
import { PushOptions } from '../common_types';
import { indent, symbols } from '../helpers';
import { MonitorSchema } from './monitor';

/* eslint-disable @typescript-eslint/no-var-requires */
const { version } = require('../../package.json');

export type APISchema = {
  project: string;
  keep_stale: boolean;
  monitors: MonitorSchema[];
};

function getAPIUrl(options: PushOptions) {
  return (
    options.url + `/s/${options.space}/api/synthetics/service/project/monitors`
  );
}

export function ok(statusCode: number) {
  return statusCode >= 200 && statusCode <= 299;
}

export async function createMonitors(
  monitors: MonitorSchema[],
  options: PushOptions
) {
  const schema: APISchema = {
    project: options.project,
    keep_stale: false,
    monitors,
  };

  return await request(getAPIUrl(options), {
    method: 'PUT',
    body: JSON.stringify(schema),
    headers: {
      authorization: `ApiKey ${options.auth}`,
      'content-type': 'application/json',
      'user-agent': `Elastic/Synthetics ${version}`,
      'kbn-xsrf': 'true',
    },
  });
}

export type APIMonitorError = {
  id?: string;
  reason: string;
  details: string;
};

export function formatNotFoundError(message: string) {
  return bold(
    `${symbols['failed']} Please check your kibana url and try again - 404:${message}`
  );
}

export function formatAPIError(
  statuCode: number,
  error: string,
  message: string
) {
  let outer = bold(`${symbols['failed']} Error\n`);
  let inner = bold(
    `${symbols['failed']} monitor creation failed - ${statuCode}:${error}\n`
  );
  inner += indent(message, '    ');
  outer += indent(inner);
  return outer;
}

export function formatFailedMonitors(errors: APIMonitorError[]) {
  let outer = bold(`${symbols['failed']} Error\n`);
  for (const error of errors) {
    const monitorId = error.id ? `: monitor(${error.id})` : '';
    let inner = bold(`> ${error.reason}${monitorId}\n`);
    inner += indent(error.details, '    ');
    outer += indent(inner) + '\n';
    outer += '\n';
  }
  return outer;
}
