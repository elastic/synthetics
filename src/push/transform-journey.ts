/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { transformFileAsync } from '@babel/core';
import type { PluginObj, types } from '@babel/core';

type JourneyPluginOptions = {
  name: string;
};

export function JourneyTransformPlugin(
  { types: t }: { types: typeof types },
  opts: JourneyPluginOptions
): PluginObj {
  return {
    name: 'transform-journeys',
    visitor: {
      CallExpression(path) {
        const { callee } = path.node;
        if (t.isIdentifier(callee) && callee.name === 'journey') {
          const args = path.node.arguments;
          if (!t.isStringLiteral(args[0])) {
            return;
          }
          if (args[0].value === opts.name) {
            path.skip();
          } else {
            path.parentPath.remove();
          }
        }
      },
    },
  };
}

export async function transform(absPath: string, journeyName: string) {
  return await transformFileAsync(absPath, {
    ast: false,
    retainLines: true,
    babelrc: false,
    configFile: false,
    plugins: [[JourneyTransformPlugin, { name: journeyName }]],
  });
}
