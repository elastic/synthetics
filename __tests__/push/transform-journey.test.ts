/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { join } from 'path';
import { transform } from '../../src/push/transform-journey';
import { mkdir, rm, writeFile } from 'fs/promises';

describe('TransformJourneyPlugin', () => {
  const PROJECT_DIR = join(__dirname, 'test-transform');
  const journeyFile = join(PROJECT_DIR, 'transform.journey.ts');

  beforeAll(async () => {
    await mkdir(PROJECT_DIR, { recursive: true });
  });

  afterAll(async () => {
    await rm(PROJECT_DIR, { recursive: true });
  });
  it('static journeys', async () => {
    await writeFile(
      journeyFile,
      `import {journey, step, monitor} from '../../';
journey('j1', () => {
monitor.use({ id: 'duplicate id' });
});

function util(){}

journey('j2', () => {});`
    );

    const res1 = await transform(journeyFile, 'j1');
    expect(res1?.code).toMatchSnapshot();
    const res2 = await transform(journeyFile, 'j2');
    expect(res2?.code).toMatchSnapshot();
    const res3 = await transform(journeyFile, '');
    expect(res3?.code).toMatchSnapshot();
  });

  it('dynamic journeys', async () => {
    await writeFile(
      journeyFile,
      `import {journey, step, monitor} from '../../';

journey('j1', () => {});

const createJourney = (name) => {
  journey(name, () => {
    monitor.use({ id: 'duplicate id' });
    step("step1", () => {})
  });
};

createJourney('j2');
`
    );

    const res1 = await transform(journeyFile, 'j1');
    expect(res1?.code).toMatchSnapshot();
    const res2 = await transform(journeyFile, 'j2');
    expect(res2?.code).toMatchSnapshot();
    const res3 = await transform(journeyFile, '');
    expect(res3?.code).toMatchSnapshot();
  });
});
