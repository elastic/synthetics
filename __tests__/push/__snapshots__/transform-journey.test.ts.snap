// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransformJourneyPlugin alias import specifier 1`] = `
"import { journey as x } from '@elastic/synthetics';

x('j1', () => {});"
`;

exports[`TransformJourneyPlugin alias import specifier 2`] = `
"import { journey as x } from '@elastic/synthetics';


x('j2', () => {});"
`;

exports[`TransformJourneyPlugin alias import specifier 3`] = `
"import { journey as x } from '@elastic/synthetics';

x('j1', () => {});
x('j2', () => {});"
`;

exports[`TransformJourneyPlugin dynamic journeys 1`] = `
"import { journey, step, monitor } from '@elastic/synthetics';

journey('j1', () => {});

const createJourney = (name) => {
  journey(name, () => {
    monitor.use({ id: 'duplicate id' });
    step("step1", () => {});
  });
};

createJourney('j2');"
`;

exports[`TransformJourneyPlugin dynamic journeys 2`] = `
"import { journey, step, monitor } from '@elastic/synthetics';



const createJourney = (name) => {
  journey(name, () => {
    monitor.use({ id: 'duplicate id' });
    step("step1", () => {});
  });
};

createJourney('j2');"
`;

exports[`TransformJourneyPlugin dynamic journeys 3`] = `
"import { journey, step, monitor } from '@elastic/synthetics';

journey('j1', () => {});

const createJourney = (name) => {
  journey(name, () => {
    monitor.use({ id: 'duplicate id' });
    step("step1", () => {});
  });
};

createJourney('j2');"
`;

exports[`TransformJourneyPlugin static journeys 1`] = `
"import { journey, step, monitor } from '@elastic/synthetics';
journey('j1', () => {
  monitor.use({ id: 'duplicate id' });
});

function util() {}"
`;

exports[`TransformJourneyPlugin static journeys 2`] = `
"import { journey, step, monitor } from '@elastic/synthetics';




function util() {}

journey('j2', () => {});"
`;

exports[`TransformJourneyPlugin static journeys 3`] = `
"import { journey, step, monitor } from '@elastic/synthetics';
journey('j1', () => {
  monitor.use({ id: 'duplicate id' });
});

function util() {}

journey('j2', () => {});"
`;
